helper-scripts/backup_and_restore.sh

```diff
@@ -1,331 +1,621 @@
+RESET="\e[0m"
 
-if [[ ! ${1} =~ (backup|restore) ]]; then
-  echo "First parameter needs to be 'backup' or 'restore'"
-  exit 1
-fi
+function print_usage() {
+  echo -e "${BOLD}Usage:${RESET} ${0} [command] [command_options]${RESET}"
+  echo
+  echo -e "${BOLD}Commands:${RESET}"
+  echo -e "\t${GREEN_COLOR}-h, --help${RESET}\t\t\t${ITALIC}Print this help message${RESET}"
+  echo -e "\t${GREEN_COLOR}-b, --backup${RESET}\t${ITALIC}<${GREEN_COLOR}path${RESET}${ITALIC}>\t\tBackup one or more components of mailcow (example: ${0} -b /path/to/backup/folder -c vmail -c crypt)${RESET}"
+  echo -e "\t${GREEN_COLOR}-r, --restore${RESET}\t${ITALIC}<${GREEN_COLOR}path${RESET}${ITALIC}>\t\tRestore the full mailcow configuration from a backup${RESET}"
+  echo -e "\t${GREEN_COLOR}-d, --delete${RESET}\t${ITALIC}<${GREEN_COLOR}path${RESET}${ITALIC}> <${RESET}${GREEN_COLOR}days${RESET}${ITALIC}>\tDelete backups older than X days in the given path${RESET}"
+  echo
+  echo -e "${BOLD}Command Options:${RESET}"
+  echo -e "\t${GREEN_COLOR}-t, --threads${RESET}\t${ITALIC}<${GREEN_COLOR}num${RESET}${ITALIC}>\t\tSet the thread count for backup and restore operations${RESET}"
+  echo -e "\t${GREEN_COLOR}-c, --component${RESET}\t${ITALIC}<${GREEN_COLOR}component${RESET}${ITALIC}>\tSet the component(s) to backup or restore [vmail, crypt, redis, rspamd, postfix, mysql and all]${RESET}"
+  echo
+}
 
-if [[ ${1} == "backup" && ! ${2} =~ (crypt|vmail|redis|rspamd|postfix|mysql|all|--delete-days) ]]; then
-  echo "Second parameter needs to be 'vmail', 'crypt', 'redis', 'rspamd', 'postfix', 'mysql', 'all' or '--delete-days'"
-  exit 1
-fi
+function check_required_tools() {
+  # Add the required tools to the array
+  local required_tools=("docker")
 
-if [[ -z ${BACKUP_LOCATION} ]]; then
-  while [[ -z ${BACKUP_LOCATION} ]]; do
-    read -ep "Backup location (absolute path, starting with /): " BACKUP_LOCATION
+  for bin in "${required_tools[@]}"; do
+    if [[ -z $(which ${bin}) ]]; then
+      echo -e "${RED_COLOR}Cannot find ${bin} in local PATH, exiting...${RESET}"
+      exit 1
+    fi
   done
-fi
 
-if [[ ! ${BACKUP_LOCATION} =~ ^/ ]]; then
-  echo "Backup directory needs to be given as absolute path (starting with /)."
-  exit 1
-fi
+  if grep --help 2>&1 | head -n 1 | grep -q -i "busybox"; then
+    echo -e "${RED_COLOR}BusyBox grep detected on local system, please install GNU grep${RESET}"
+    exit 1
+  fi
+}
 
-if [[ -f ${BACKUP_LOCATION} ]]; then
-  echo "${BACKUP_LOCATION} is a file!"
-  exit 1
-fi
+# declare_restore_point declares the `RESTORE_POINT`
+# globally which is the path to the backup folder.
+#
+# If the function succeeds, the variable `RESTORE_POINT`
+# will be declared globally.
+# If the function fails, it will exit with a status of 1.
+function declare_restore_point() {
+  local BACKUP_LOCATION="${1}"
+
+  # Check subfolders inside BACKUP_LOCATION
+  if [[ $(find "${BACKUP_LOCATION}"/mailcow-* -maxdepth 1 -type d 2> /dev/null| wc -l) -lt 1 ]]; then
```

Comment:
I don't know for sure, if find has a similar behavior like grep on busybox or if it is the same for both?

Reply:
Yes, searched for it. It has BusyBox version too. Added to the check list for BusyBox.